networkit/cpp/io/ThrillGraphBinaryWriter.cpp

```diff
@@ -0,0 +1,115 @@
+		s << uint8_t(((v >> 07) & 0x7F) | 0x80);
+		s << uint8_t(((v >> 14) & 0x7F) | 0x80);
+		s << uint8_t(((v >> 21) & 0x7F) | 0x80);
+		s << uint8_t((v >> 28) & 0x7F);
+	}
+	else if (v < 128llu * 128 * 128 * 128 * 128 * 128) {
+		s << uint8_t(((v >> 00) & 0x7F) | 0x80);
+		s << uint8_t(((v >> 07) & 0x7F) | 0x80);
+		s << uint8_t(((v >> 14) & 0x7F) | 0x80);
+		s << uint8_t(((v >> 21) & 0x7F) | 0x80);
+		s << uint8_t(((v >> 28) & 0x7F) | 0x80);
+		s << uint8_t((v >> 35) & 0x7F);
+	}
+	else if (v < 128llu * 128 * 128 * 128 * 128 * 128 * 128) {
+		s << uint8_t(((v >> 00) & 0x7F) | 0x80);
+		s << uint8_t(((v >> 07) & 0x7F) | 0x80);
+		s << uint8_t(((v >> 14) & 0x7F) | 0x80);
+		s << uint8_t(((v >> 21) & 0x7F) | 0x80);
+		s << uint8_t(((v >> 28) & 0x7F) | 0x80);
+		s << uint8_t(((v >> 35) & 0x7F) | 0x80);
+		s << uint8_t((v >> 42) & 0x7F);
+	}
+	else if (v < 128llu * 128 * 128 * 128 * 128 * 128 * 128 * 128) {
+		s << uint8_t(((v >> 00) & 0x7F) | 0x80);
+		s << uint8_t(((v >> 07) & 0x7F) | 0x80);
+		s << uint8_t(((v >> 14) & 0x7F) | 0x80);
+		s << uint8_t(((v >> 21) & 0x7F) | 0x80);
+		s << uint8_t(((v >> 28) & 0x7F) | 0x80);
+		s << uint8_t(((v >> 35) & 0x7F) | 0x80);
+		s << uint8_t(((v >> 42) & 0x7F) | 0x80);
+		s << uint8_t((v >> 49) & 0x7F);
+	}
+	else if (v < 128llu * 128 * 128 * 128 * 128 * 128 * 128 * 128 * 128) {
+		s << uint8_t(((v >> 00) & 0x7F) | 0x80);
+		s << uint8_t(((v >> 07) & 0x7F) | 0x80);
+		s << uint8_t(((v >> 14) & 0x7F) | 0x80);
+		s << uint8_t(((v >> 21) & 0x7F) | 0x80);
+		s << uint8_t(((v >> 28) & 0x7F) | 0x80);
+		s << uint8_t(((v >> 35) & 0x7F) | 0x80);
+		s << uint8_t(((v >> 42) & 0x7F) | 0x80);
+		s << uint8_t(((v >> 49) & 0x7F) | 0x80);
+		s << uint8_t((v >> 56) & 0x7F);
+	}
+	else {
+		s << uint8_t(((v >> 00) & 0x7F) | 0x80);
+		s << uint8_t(((v >> 07) & 0x7F) | 0x80);
+		s << uint8_t(((v >> 14) & 0x7F) | 0x80);
+		s << uint8_t(((v >> 21) & 0x7F) | 0x80);
+		s << uint8_t(((v >> 28) & 0x7F) | 0x80);
+		s << uint8_t(((v >> 35) & 0x7F) | 0x80);
+		s << uint8_t(((v >> 42) & 0x7F) | 0x80);
+		s << uint8_t(((v >> 49) & 0x7F) | 0x80);
+		s << uint8_t(((v >> 56) & 0x7F) | 0x80);
+		s << uint8_t((v >> 63) & 0x7F);
+	}
+
+	return s;
+};
+
+
+}
+
+
+void NetworKit::ThrillGraphBinaryWriter::write( const NetworKit::Graph &G, const std::string &path ) {
+	if (G.numberOfNodes() > std::numeric_limits<uint32_t>::max()) {
+		throw std::runtime_error("Thrill binary graphs only support graphs with up to 2^32-1 nodes.");
+	}
+	
+	std::ofstream out_stream(path, std::ios::trunc | std::ios::binary);
+
+	std::vector<uint32_t> neighbors;
+	G.forNodes([&](node u) {
+			neighbors.clear();
+			G.forEdgesOf(u, [&](node v) {
+					if (u <= v) {
+						neighbors.push_back(v);
+					}
+			});
+
+			PutVarint(out_stream, neighbors.size());
+			
+			for (uint32_t v : neighbors) {
+				static_assert(std::is_same<uint32_t, decltype(v)>::value, "Node type is not uint32 anymore, adjust code!");
+				out_stream.write(reinterpret_cast<const char*>(&v), 4);
```

Comment:
Okay, we write here in host byte order which means that the readers (that I commented on) work correctly on any machine. However, transferring the files from little endian to big endian or vice-versa is not possible. Is that intended?

Reply:
I was aware of that, but as far as I understand the Thrill code, it is doing the same (i.e., uses the host's endianness). Nevertheless I think it would be a good idea to fix the endianness, so a7ff6f55733400afb172d0a24669715cb7fea8d2 now reads individual bytes. I chose little endian because it is what will be used on like most current compute clusters where you execute Thrill or NetworKit and is thus also the format my current graph files use.